release-22.2: upgrades: make upgrade to sequence by ID better

ajwerner opened the pull request:
This makes the upgrade robust to dropped descriptors, and makes it avoid performing a write to already upgraded descriptors.

Epic: none

fixes #99864
Release justification: upgrade improvement
Release note (bug fix): Fixed a bug which could cause an upgrade from 22.1 to 22.2 to take a long time and sometimes not finish if there are a lot of tables and they are being created and dropped regularly.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99936)
<!-- Reviewable:end -->

ajwerner commented:
@Xiang-Gu can you review this and pick it up by writing tests. A user out there has been stuck on this migration.

Xiang-Gu committed:
release-22.2: upgrades: make upgrade to sequence by ID better

This makes the upgrade robust to dropped descriptors, and makes it avoid performing a write to already upgraded descriptors.

Epic: none

Relates to #99864

Release note (bug fix): Fixed a bug which could cause an upgrade
from 22.1 to 22.2 to take a long time and sometimes not finish if there
are a lot of tables and they are being created and dropped regularly.

rafiss reviewed (approved):
thank you!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99936#-:-NYyr_ng04GctkpSXKoB:bsv417h)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
